File: src/mk3/JavaPackage/src/main/java/tamaf/Behavior.java
package tamaf;

import tamaf.defines.EmptyReceiveException;
import java.util.ArrayList;
import java.util.List;

public abstract class Behavior {
    protected Agent agent;
    private boolean isBlocked = false;
    private boolean isStarted = false;
    protected List<ACLMessage> messagesToPutBack = new ArrayList<>();
    protected EMAInteraction emaInteraction;

    public Behavior() {
        this.emaInteraction = new EMAInteraction(this);
    }
    public Behavior(Agent agent) {
        this.agent = agent;
        this.emaInteraction = new EMAInteraction(this);
    }

    public void setAgent(Agent agent) { this.agent = agent; }
    public Agent getAgent() { return agent; }

    public void onStart() {}
    public abstract void action();
    public abstract boolean done();
    public int onEnd() { return 0; }

    public void block() { this.isBlocked = true; }
    public void unblock() { this.isBlocked = false; }
    public boolean isBlocked() { return isBlocked; }

    public void setStarted(boolean started) { this.isStarted = started; }
    public boolean isStarted() { return isStarted; }

    public ACLMessage receive(ACLMessageTemplate template) {
        ACLMessage msg = this.agent.receive(template);
        if (msg == null) {
            throw new EmptyReceiveException();
        } else {
            messagesToPutBack.add(0, msg);
        }
        return msg;
    }

    public void execute() {
        messagesToPutBack.clear();
        try {
            this.unblock();
            this.action();
        } catch (EmptyReceiveException e) {
            this.block();
            for (ACLMessage msg : messagesToPutBack) {
                this.putBack(msg);
            }
        }
    }

    public boolean send(ACLMessage msg) {
        return this.agent.send(msg);
    }

    public void putBack(ACLMessage msg) {
        this.agent.putBack(msg);
    }
}

File: src/mk3/PythonPackage/src/tamaf/behavior.py
from abc import ABC, abstractmethod
from typing import TYPE_CHECKING
from .defines import EmptyReceiveException

if TYPE_CHECKING:
    from .aclmessage import ACLMessage
    from .aclmessagetemplate import ACLMessageTemplate
    from .agent import Agent
    from .emainteraction import EMAInteraction

class Behavior(ABC):
    
    def __init__(self, agent: 'Agent' = None):
        from .emainteraction import EMAInteraction
        self.isBlocked = False
        self.isStarted = False
        self.agent = agent
        self.messagesToPutBack: list['ACLMessage'] = []
        self.emaInteraction = EMAInteraction(self)

    def OnStart(self):
        pass
    
    @abstractmethod
    def Action(self):
        pass

    @abstractmethod
    def Done(self):
        pass

    def OnEnd(self):
        return 0

    def Block(self):
        self.isBlocked = True
    
    def UnBlock(self):
        self.isBlocked = False

    def IsBlocked(self):
        return self.isBlocked

    def Started(self):
        self.isStarted = True
    
    def IsStarted(self):
        return self.isStarted
    
    def Receive(self, aclMessageTemplate : 'ACLMessageTemplate'):
        aclMessage = self.agent.Receive(aclMessageTemplate)

        if aclMessage is None:
            raise EmptyReceiveException
        else:
            self.messagesToPutBack.insert(0, aclMessage)
        
        return aclMessage

    def Execute(self):
        self.messagesToPutBack: list['ACLMessage'] = []

        self.UnBlock()
        
        try:
            self.Action()

        except EmptyReceiveException:
            self.Block()
            for aclMessage in self.messagesToPutBack:
                self.PutBack(aclMessage)

    def Send(self, aclMessage: 'ACLMessage'):
        return self.agent.Send(aclMessage)
    
    def PutBack(self, aclMessage: 'ACLMessage'):
            return self.agent.PutBack(aclMessage)
